File: wrappers/javascript/shared/src/store/Session.ts
/* eslint-disable @typescript-eslint/no-unsafe-return */
import type { Key, SessionHandle, StoreHandle } from '../crypto'
import type { KeyAlgs } from '../enums'

import { ariesAskar } from '../ariesAskar'
import { EntryOperation } from '../enums/EntryOperation'
import { AriesAskarError } from '../error'

import { Entry } from './Entry'
import { EntryList } from './EntryList'
import { KeyEntryList } from './KeyEntryList'

export class Session {
  // TODO: where is the store used?
  private store: StoreHandle
  private _handle?: SessionHandle
  private isTxn: boolean

  public constructor({ store, handle, isTxn }: { store: StoreHandle; handle?: SessionHandle; isTxn: boolean }) {
    this.store = store
    this._handle = handle
    this.isTxn = isTxn
  }

  public get isTransaction() {
    return this.isTxn
  }

  public get handle() {
    return this._handle
  }

  public async count({ category, tagFilter }: { category: string; tagFilter?: Record<string, unknown> }) {
    if (!this.handle) throw new AriesAskarError({ code: 100, message: 'Cannot count from closed session' })
    return await ariesAskar.sessionCount({ tagFilter, category, sessionHandle: this.handle })
  }

  public async fetch({
    category,
    name,
    forUpdate = false,
    isJson = false,
  }: {
    category: string
    name: string
    forUpdate?: boolean
    isJson?: boolean
  }) {
    if (!this.handle) throw new AriesAskarError({ code: 100, message: 'Cannot fetch from a closed session' })

    const handle = await ariesAskar.sessionFetch({ forUpdate, name, category, sessionHandle: this.handle })
    if (!handle) return undefined

    const entry = new Entry({ list: handle, pos: 0 })
    // console.log(entry)

    return entry.toJson(isJson)
  }

  public async fetchAll({
    category,
    forUpdate = false,
    limit,
    tagFilter,
  }: {
    category: string
    tagFilter?: Record<string, unknown>
    limit?: number
    forUpdate?: boolean
  }) {
    if (!this.handle) throw new AriesAskarError({ code: 100, message: 'Cannot fetch all from a closed session' })
    const handle = await ariesAskar.sessionFetchAll({
      forUpdate,
      limit,
      tagFilter,
      sessionHandle: this.handle,
      category,
    })
    return new EntryList({ handle })
  }

  public async insert({
    category,
    name,
    expiryMs,
    tags,
    value,
  }: {
    category: string
    name: string
    value: string | Record<string, unknown>
    tags?: Record<string, unknown>
    expiryMs?: number
  }) {
    if (!this.handle) throw new AriesAskarError({ code: 100, message: 'Cannot insert with a closed session' })
    const serializedValue = JSON.stringify(value)
    if (!serializedValue)
      throw new AriesAskarError({ code: 100, message: 'Either `value` or `valueJson` must be defined' })

    // @ts-ignore
    const encoder = new TextEncoder()

    await ariesAskar.sessionUpdate({
      // @ts-ignore
      value: new Uint8Array(encoder.encode(serializedValue)),
      expiryMs,
      tags,
      name,
      category,
      sessionHandle: this.handle,
      operation: EntryOperation.Insert,
    })
  }

  public async replace({
    category,
    name,
    expiryMs,
    tags,
    value,
    valueJson,
  }: {
    category: string
    name: string
    value?: string
    tags?: Record<string, unknown>
    expiryMs?: number
    valueJson?: unknown
  }) {
    if (!this.handle) throw new AriesAskarError({ code: 100, message: 'Cannot replace with a closed session' })
    const serializedValue = !value && valueJson ? JSON.stringify(valueJson) : value
    if (!serializedValue)
      throw new AriesAskarError({ code: 100, message: 'Either `value` or `valueJson` must be defined' })

    // @ts-ignore
    const encoder = new TextEncoder()

    await ariesAskar.sessionUpdate({
      // @ts-ignore
      value: new Uint8Array(encoder.encode(serializedValue)),
      expiryMs,
      tags,
      name,
      category,
      sessionHandle: this.handle,
      operation: EntryOperation.Replace,
    })
  }

  public async remove({ category, name }: { category: string; name: string }) {
    if (!this.handle) throw new AriesAskarError({ code: 100, message: 'Cannot remove with a closed session' })

    await ariesAskar.sessionUpdate({
      name,
      category,
      sessionHandle: this.handle,
      operation: EntryOperation.Remove,
    })
  }

  public async removeAll({ category, tagFilter }: { category: string; tagFilter: Record<string, unknown> }) {
    if (!this.handle) throw new AriesAskarError({ code: 100, message: 'Cannot remove all with a closed session' })

    await ariesAskar.sessionRemoveAll({
      category,
      sessionHandle: this.handle,
      tagFilter,
    })
  }

  public async insertKey({
    name,
    key,
    expiryMs,
    metadata,
    tags,
  }: {
    name: string
    key: Key
    metadata?: string
    tags?: string
    expiryMs?: number
  }) {
    if (!this.handle) throw new AriesAskarError({ code: 100, message: 'Cannot insert a key with a closed session' })

    await ariesAskar.sessionInsertKey({
      expiryMs,
      tags,
      metadata,
      name,
      sessionHandle: this.handle,
      localKeyHandle: key.handle,
    })
  }

  public async fetchKey({ name, forUpdate }: { name: string; forUpdate: boolean }) {
    if (!this.handle) throw new AriesAskarError({ code: 100, message: 'Cannot fetch a key with a closed session' })
    const handle = await ariesAskar.sessionFetchKey({ forUpdate, name, sessionHandle: this.handle })
    //TODO: what to return here
  }

  public async fetchAllKeys({
    forUpdate,
    alg,
    limit,
    tagFilter,
    thumbprint,
  }: {
    alg?: KeyAlgs
    thumbprint?: string
    tagFilter?: Record<string, unknown>
    limit?: number
    forUpdate: boolean
  }) {
    if (!this.handle) throw new AriesAskarError({ code: 100, message: 'Cannot fetch all keys with a closed session' })
    const handle = await ariesAskar.sessionFetchAllKeys({
      forUpdate,
      limit,
      tagFilter,
      thumbprint,
      alg,
      sessionHandle: this.handle,
    })

    return new KeyEntryList({ handle })
  }

  public async updateKey({
    name,
    expiryMs,
    metadata,
    tags,
  }: {
    name: string
    metadata?: string
    tags?: Record<string, unknown>
    expiryMs?: number
  }) {
    if (!this.handle) throw new AriesAskarError({ code: 100, message: 'Cannot update a key with a closed session' })
    await ariesAskar.sessionUpdateKey({ expiryMs, tags, metadata, name, sessionHandle: this.handle })
  }

  public async removeKey({ name }: { name: string }) {
    if (!this.handle) throw new AriesAskarError({ code: 100, message: 'Cannot remove a key with a closed session' })
    await ariesAskar.sessionRemoveKey({ name, sessionHandle: this.handle })
  }

  public async commit() {
    if (!this.isTxn) throw new AriesAskarError({ code: 100, message: 'Session is not a transaction' })
    if (!this.handle) throw new AriesAskarError({ code: 100, message: 'Cannot commit a closed session' })
    await this.handle.close(true)
    this._handle = undefined
  }

  public async rollback() {
    if (!this.isTxn) throw new AriesAskarError({ code: 100, message: 'Session is not a transaction' })
    if (!this.handle) throw new AriesAskarError({ code: 100, message: 'Cannot rollback a closed session' })
    await this.handle.close(false)
    this._handle = undefined
  }

  public async close() {
    if (!this.handle) throw new AriesAskarError({ code: 100, message: 'Cannot close a closed session' })
    await this.handle.close(false)
    this._handle = undefined
  }
}

File: wrappers/swift/Askar/Sources/Askar/Session.swift
import Foundation
import AskarFramework

public class Session {
    let _handle: SessionHandle
    private static var continuation: CheckedContinuation<Any, Error>?

    public init(handle: SessionHandle) {
        self._handle = handle
    }

    public func count(category: String, tagFilter: String) async throws -> Int {
        let count = try await withCheckedThrowingContinuation { (continuation: CheckedContinuation<Any, Error>) in
            Session.continuation = continuation
            let err = askar_session_count(_handle, category, tagFilter, { (_, err, count) in
                if err != Success {
                    Session.continuation?.resume(throwing: AskarError.nativeError(code: err.rawValue))
                } else {
                    Session.continuation?.resume(returning: count)
                }
            }, 0)
            if err != Success {
                Session.continuation?.resume(throwing: AskarError.nativeError(code: err.rawValue))
            }
        } as! Int64

        return Int(count)
    }

    public func fetch(category: String, name: String, forUpdate: Bool = false) async throws -> Entry? {
        let handle = try await withCheckedThrowingContinuation { (continuation: CheckedContinuation<Any, Error>) in
            Session.continuation = continuation
            let err = askar_session_fetch(_handle, category, name, forUpdate ? 1:0, { (_, err, handle) in
                if err != Success {
                    Session.continuation?.resume(throwing: AskarError.nativeError(code: err.rawValue))
                } else {
                    Session.continuation?.resume(returning: handle)
                }
            }, 0)
            if err != Success {
                Session.continuation?.resume(throwing: AskarError.nativeError(code: err.rawValue))
            }
        } as! EntryListHandle

        if !handle.isEmpty {
            return try EntryList(handle: handle, len: 1).next()
        } else {
            return nil
        }
    }

    public func fetchAll(category: String, tagFilter: String? = nil, limit: Int = -1, forUpdate: Bool = false) async throws -> EntryList? {
        let handle = try await withCheckedThrowingContinuation { (continuation: CheckedContinuation<Any, Error>) in
            Session.continuation = continuation
            let err = askar_session_fetch_all(_handle, category, tagFilter, Int64(limit), forUpdate ? 1:0, { (_, err, handle) in
                if err != Success {
                    Session.continuation?.resume(throwing: AskarError.nativeError(code: err.rawValue))
                } else {
                    Session.continuation?.resume(returning: handle)
                }
            }, 0)
            if err != Success {
                Session.continuation?.resume(throwing: AskarError.nativeError(code: err.rawValue))
            }
        } as! EntryListHandle

        if !handle.isEmpty {
            return try EntryList(handle: handle)
        } else {
            return nil
        }
    }

    public func update(operation: EntryOperation, category: String, name: String, value: Data? = nil, tags: String? = nil, expiryMillis: Int = -1) async throws {
        let byteBuf = FfiByteBuffer(fromData: value)
        _ = try await withCheckedThrowingContinuation { (continuation: CheckedContinuation<Any, Error>) in
            Session.continuation = continuation
            let err = askar_session_update(_handle, operation.rawValue, category, name, byteBuf.buffer, tags, Int64(expiryMillis), { (_, err) in
                if err != Success {
                    Session.continuation?.resume(throwing: AskarError.nativeError(code: err.rawValue))
                } else {
                    Session.continuation?.resume(returning: ())
                }
            }, 0)
            if err != Success {
                Session.continuation?.resume(throwing: AskarError.nativeError(code: err.rawValue))
            }
        }
    }

    public func insertKey(_ key: Key, name: String, meta: String? = nil, tags: String? = nil, expiryMillis: Int = -1) async throws {
        _ = try await withCheckedThrowingContinuation { (continuation: CheckedContinuation<Any, Error>) in
            Session.continuation = continuation
            let err = askar_session_insert_key(_handle, key.handle, name, meta, tags, Int64(expiryMillis), { (_, err) in
                if err != Success {
                    Session.continuation?.resume(throwing: AskarError.nativeError(code: err.rawValue))
                } else {
                    Session.continuation?.resume(returning: ())
                }
            }, 0)
            if err != Success {
                Session.continuation?.resume(throwing: AskarError.nativeError(code: err.rawValue))
            }
        }
    }

    public func fetchKey(name: String, forUpdate: Bool = false) async throws -> KeyEntry? {
        let handle = try await withCheckedThrowingContinuation { (continuation: CheckedContinuation<Any, Error>) in
            Session.continuation = continuation
            let err = askar_session_fetch_key(_handle, name, forUpdate ? 1:0, { (_, err, handle) in
                if err != Success {
                    Session.continuation?.resume(throwing: AskarError.nativeError(code: err.rawValue))
                } else {
                    Session.continuation?.resume(returning: handle)
                }
            }, 0)
            if err != Success {
                Session.continuation?.resume(throwing: AskarError.nativeError(code: err.rawValue))
            }
        } as! KeyEntryListHandle

        if !handle.isEmpty {
            return try KeyEntryList(handle: handle, len: 1).next()
        } else {
            return nil
        }
    }

    public func fetchAllKeys(alg: KeyAlg? = nil, thumbprint: String? = nil, tagFilter: String? = nil, limit: Int = -1, forUpdate: Bool = false) async throws -> KeyEntryList? {
        let handle = try await withCheckedThrowingContinuation { (continuation: CheckedContinuation<Any, Error>) in
            Session.continuation = continuation
            let err = askar_session_fetch_all_keys(_handle, alg?.rawValue, thumbprint, tagFilter, Int64(limit), forUpdate ? 1:0, { (_, err, handle) in
                if err != Success {
                    Session.continuation?.resume(throwing: AskarError.nativeError(code: err.rawValue))
                } else {
                    Session.continuation?.resume(returning: handle)
                }
            }, 0)
            if err != Success {
                Session.continuation?.resume(throwing: AskarError.nativeError(code: err.rawValue))
            }
        } as! KeyEntryListHandle

        if !handle.isEmpty {
            return try KeyEntryList(handle: handle)
        } else {
            return nil
        }
    }

    public func updateKey(name: String, meta: String? = nil, tags: String? = nil, expiryMillis: Int = -1) async throws {
        _ = try await withCheckedThrowingContinuation { (continuation: CheckedContinuation<Any, Error>) in
            Session.continuation = continuation
            let err = askar_session_update_key(_handle, name, meta, tags, Int64(expiryMillis), { (_, err) in
                if err != Success {
                    Session.continuation?.resume(throwing: AskarError.nativeError(code: err.rawValue))
                } else {
                    Session.continuation?.resume(returning: ())
                }
            }, 0)
            if err != Success {
                Session.continuation?.resume(throwing: AskarError.nativeError(code: err.rawValue))
            }
        }
    }

    public func removeKey(name: String) async throws {
        _ = try await withCheckedThrowingContinuation { (continuation: CheckedContinuation<Any, Error>) in
            Session.continuation = continuation
            let err = askar_session_remove_key(_handle, name, { (_, err) in
                if err != Success {
                    Session.continuation?.resume(throwing: AskarError.nativeError(code: err.rawValue))
                } else {
                    Session.continuation?.resume(returning: ())
                }
            }, 0)
            if err != Success {
                Session.continuation?.resume(throwing: AskarError.nativeError(code: err.rawValue))
            }
        }
    }

    public func close() async throws {
        _ = try await withCheckedThrowingContinuation { (continuation: CheckedContinuation<Any, Error>) in
            Session.continuation = continuation
            let err = askar_session_close(_handle, 0, { (_, err) in
                if err != Success {
                    Session.continuation?.resume(throwing: AskarError.nativeError(code: err.rawValue))
                } else {
                    Session.continuation?.resume(returning: ())
                }
            }, 0)
            if err != Success {
                Session.continuation?.resume(throwing: AskarError.nativeError(code: err.rawValue))
            }
        }
    }
}
